### PR TITLE
Added NaNReporter and HighMaReporter to LettuceCFD

### LINKED ISSUE #174: Reporter to interrupt simulation
Often, the server time is limited on the cluster. A reporter should be able to check the expired time and interrupt Simulation.__call__() by simply setting flow.i to num_steps if the time surpasses a limit.

This could be useful for the benchmark, see #92

### LINKED ISSUE #272: [Feature]: abort conditions - there are several reasons a simulation can fail
### Contact Details

_No response_

### Is your proposal related to a problem?

Simulations can fail due to several reasons:

- NaN: when NaN values occur, the simulation is crashed. It should probably be aborted and the crash can be reported. The NaN reporter of #248 does that.
- time constraints: your maximum simulation wall time may be limited (for example when you are using a HPC cluster and a maximum job time is 24 h). When the scheduler kills your job, all data could be lost.
- other reasons: whatever you can think of

It would be nice if the program had several abort conditions, giving messages and information about the reason, time and circumstances of simulation termination. Also: termination could be wanted (NaN encountered, or close to time limit). Data up until this point could be saved (simulation state checkpoint and data export to files).

### Describe the solution you'd like

It would be nice to have a couple of simulation abort conditions, which simulation-class or a reporter checks from time to time and aborts the simulation, if a condition is met and treats remaining data and sim.-state as needed.

example conditions:
0=fine, 1=timelimit reached, 2=nan found, 3=Ma>0.3 (and choose if sim. should stop or just report that),...

### Describe alternatives you've considered

The implementation depends on who handles what: simulation, reporter or something else.
1. Set simulation status for abort condition (reporter?)
2. check if abort condition (simulation?)
3. process data: write checkpoint and/or export data as needed (reporter or simulation?)
4. end simulation and report abort condition and circumstances

### Additional context

One could think about more features of the checkpointing functionality of simulaiton -> new issue...;
Torch doesn't like it, when the device on which the imported date was saved from (e.g. "cuda:1") is not the same as the device which lettuce is currently working on (e.g. "cuda:0").

Example: simulation 1 runs on cpu, a checkpoint is written...; later, simulation 2, which runs on gpu, should import that checkpoint as an initial condition. This would create a problem. 
Another example: depending on the device-association on multi-GPU-computers, a similar issue might arise.

Always transferring data to the cpu before writing a checkpoint and then transferring it to the device currently in use, after reading a checkpoint can circumvent this.

### PR DESCRIPTION
**Pull request by MaxBille**
## Description

Closes #174 
Closes #272 
Adding two reporters, that check for signs of simulation crashes (NaN in f or Ma > 0.3 in u) and abort simulation with respective abort messages, optional final vtk-frame and (in case of HighMaReporter) optional output file with locations of the 100 highest velocity magnitudes for crash-troubleshooting and analysis.

NaNReporter:
- checks f in interval steps for any NaN value. If NaN is found, simulation is aborted.
- if outdir is provided, lists indices q,x,y,[z] for all found NaN values (attention: produces large files for large intervals, because NaN replicates through the domain fast.
- "old" argument (True, False) is for backwards compatibility with lettuce that has no aborting functionality in the simulation.py class. No vtk output in this case...
- vtk argument (True, False) allows for vtk output of the current single flow state frame to outdir (if vtk_dir is not specified) or vtk_dir (if specified)

HighMaReporter:
- calculates mach number from norm(u(f)) every interval steps and checks if Ma > 0.3 occurs. If Ma > 0.3 is found, simulation is aborted. The location (in LU coordinates) of up to 100 highest Ma numbers in the domain are reported to a file in outdir, if outdir argument is provided.
- "old" argument (True, False) is for backwards compatibility with lettuce that has no aborting functionality in the simulation.py class. No vtk output in this case...
- vtk argument (True, False) allows for vtk output of the current single flow state frame to outdir (if vtk_dir is not specified) or vtk_dir (if specified)
- relies on unravel_index() method for torch.tensors, defined in the same file, for sorting and selection of high-Ma-locations.

DISCLAIMER: originally written for a modified version of lettuceCFD, based on Version 0.2.3; Adaptation for new Lettuce Master pending...

## Checklist

 - [x] This pull request is associated to an issue
 - [x] This PR contains a description
 - [x] Add someone else as reviewer and wait for approval before merging.

### CODE REVIEW

**Review by PhiSpel — CHANGES_REQUESTED** [review]
1. Could you check if we need to set non_native in BreakableSimulation? If so, can you rename to `BreakableSimulationNonNative`?
2. Maybe rename example to `FailingTGVandObstacle.py`?

File: lettuce/ext/_reporter/nan_repoter.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+            locations.append("y")
+        if simulation.flow.stencil.d > 2:
+            locations.append("z")
+        return locations
+
+    def show_max(self, my_file, simulation: 'Simulation'):
+        pass
+
+
+def unravel_index(indices: torch.Tensor, shape: tuple[int, ...],
```
Comment (PhiSpel) [inline lettuce/ext/_reporter/nan_repoter.py]: Why do we need this here? Is this not included in torch? https://docs.pytorch.org/docs/stable/generated/torch.unravel_index.html

File: lettuce/ext/_reporter/nan_repoter.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+        u = simulation.flow.u()
+        ma = (torch.norm(u, dim=0)
+              / simulation.flow.stencil.cs)[self.fails(simulation)]
+        ma = simulation.context.convert_to_ndarray(ma)
+        failed_locations_list.append(ma)
+        if len(failed_locations_list[0]) >= 100:
+            return np.stack(failed_locations_list, axis=-1)[:100]
+        return np.stack(failed_locations_list, axis=-1)
+
+    def first_100(self, simulation: 'Simulation'):
```
Comment (PhiSpel) [inline lettuce/ext/_reporter/nan_repoter.py]: Is this ever applied?

File: lettuce/ext/_reporter/vtk_reporter.py [hunk 3]
```diff
@@ -4,7 +4,7 @@
 
 from ... import Reporter
 
-__all__ = ['VTKReporter']
+__all__ = ['VTKReporter', 'write_vtk']
```
Comment (PhiSpel) [inline lettuce/ext/_reporter/vtk_reporter.py]: Should we make this a public function? Why not keep it guarded behind `VTKReporter`?

File: examples/advanced_flows/FailingTGVandObstacle.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+nan_reporter = lt.NaNReporter(100, outdir="./data/nan_reporter", vtk_out=True)
+simulation = lt.BreakableSimulation(
+    flow=flow,
+    collision=lt.BGKCollision(tau=flow.units.relaxation_parameter_lu),
+    reporter=[nan_reporter])
+simulation(10000)
+print(f"Failed after {nan_reporter.failed_iteration} iterations")
+
+############################################
+# b) unstable obstacle flow,m that causes high Ma values (Ma > 0.3), which are
```
Comment (PhiSpel) [inline examples/advanced_flows/FailingTGVandObstacle.py]: typo

File: lettuce/ext/_reporter/__init__.py [hunk 5]
```diff
@@ -2,3 +2,5 @@
 from .observable_reporter import *
 from .vtk_reporter import *
 from .write_image import *
+#from .nan_repoter import *
```
Comment (PhiSpel) [inline lettuce/ext/_reporter/__init__.py]: remove redundant

File: lettuce/ext/_reporter/failure_reporter.py [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+import os
+from typing import List
+
+from ... import Reporter, BreakableSimulation
+from .vtk_reporter import write_vtk
+
+__all__ = ["FailureReporterBase", "NaNReporter", "HighMaReporter"]
+
+# Angenommene Superklasse (falls sie nicht importiert werden kann)
+# class Reporter: ...
```
Comment (PhiSpel) [inline lettuce/ext/_reporter/failure_reporter.py]: remove (?)

File: lettuce/ext/_reporter/nan_repoter.py [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+    # nan_location. This will not impact the abortion of sim. by NaN_Reporter
+
+    def __init__(self, interval=100, outdir=None, vtk=False):
+        self.outdir = outdir
+        self.vtk = vtk
+        self.name = 'NaN'  # what is the name for?
+        self.failed_iteration = None
+        super().__init__(interval)
+
+    def __call__(self, simulation: 'Simulation'):
```
Comment (PhiSpel) [inline lettuce/ext/_reporter/nan_repoter.py]: Since you import `Simulation` with no apparent issues, you can type it directly. Keep in mind, though: This must be a `BreakableSimulation`

File: lettuce/ext/_reporter/nan_repoter.py [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+                    self.outputs(simulation)
+
+                print(
+                    f'(!) ABORT MESSAGE: FailReporter detected {self.name}'
+                    f' in f (interval = {self.interval}) at iteration '
+                    f'{simulation.flow.i}. See log in {self.outdir} for '
+                    f'details!')
+                # telling simulation to abort simulation by setting i too high
+                simulation.flow.i = int(simulation.flow.i + 1e10)
+                # TODO: maybe make this more robust with a failed-flag in simulation and not rely on flow.i to be high/low?
```
[the remark this answers is not in the document]
Comment (MaxBille) [reply lettuce/ext/_reporter/nan_repoter.py]: will do

File: lettuce/ext/_reporter/nan_repoter.py (hunk 7 above)
Comment (PhiSpel) [reply lettuce/ext/_reporter/nan_repoter.py]: @MaxBille have you tried this?
